File: cs/competitive_programming/leetcode_interns/LinkedList/EASY/21_merge_two_sorted_lists.py
#!/usr/bin/python3

# Link - https://leetcode.com/problems/merge-two-sorted-lists/

from typing import List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# ITERATIVE Solution
# Time complexity : O(n + m)
# Space complexity : O(1)
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        prevHead = ListNode(-1)
        
        prev = prevHead
        
        while(l1 and l2):
            
            if l1.val <= l2.val:
                prev.next = l1
                l1 = l1.next
            else:
                prev.next = l2
                l2 = l2.next
            
            prev = prev.next
            
        # At least one of l1 and l2 can still have nodes at this point, so connect
        # the non-null list to the end of the merged list.
        
        prev.next = l1 if l1 != None else l2
        
        return prevHead.next

# RECURSIVE Solution
# Time Complexity - O(n+m)
# Space Complexity - O(n+m)

class Solution1:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        if l1 == None:
            return l2
        elif l2 == None:
            return l1
        elif l1.val <= l2.val:
            l1.next = self.mergeTwoLists(l1.next, l2)
            return l1
        else:
            l2.next = self.mergeTwoLists(l1, l2.next)
            return l2
        
        
    

File: cs/competitive_programming/leetcode_fulltimes/2023/may/21_merge_two_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        ListNode* dummy_node = new ListNode();
        ListNode* res = dummy_node;


        while(list1!=nullptr && list2!=nullptr) {

            if(list1->val<=list2->val){
                dummy_node->next = list1;
                list1=list1->next;
            } else {
                dummy_node->next = list2;
                list2=list2->next;
            }
            dummy_node = dummy_node->next;
        }
        if(list1){
            dummy_node->next = list1;
        }
        if(list2){
            dummy_node->next = list2;
        }
        return res->next;
        
    }
};
